Sources/NIO/ChannelOption.swift

```diff
@@ -327,3 +327,39 @@ extension ChannelOptions {
         }
     }
 }
+
+// Provides convenience methods for `ChannelOption.socket`
+extension ChannelOptions {
+    
+    public struct Socket {
+        
+        /// Turn on debugging info recording
+        public static let enableDebugging = ChannelOptions.Types.SocketOption(level: SOL_SOCKET, name: SO_DEBUG)
+        
+        /// Socket has had listen()
+        public static let acceptConnections = ChannelOptions.Types.SocketOption(level: SOL_SOCKET, name: SO_ACCEPTCONN)
```

Comment:
No-one ever sets this so let's skip that.

Reply:
Change made, however what would be the reasoning for not supporting a simple use case, no matter how rare?

Reply:
Lots of reasons, but here are a few:

- all code imposes maintenance costs. By default, my heuristic for writing any line of code begins with "don't" unless there is a use-case served by that code. Once written, we must support that code for the foreseeable future: code we didn't write is therefore code we don't have to support, and much cheaper than the code we did write.
- code no-one uses is silently broken. If no-one is using some code we wrote, the odds of it silently breaking or, worse, being or becoming useless are quite high. This is a double-whammy on top of the problem of having the maintenance burden of the code.
- extra code that exercises an API makes it harder to evolve that API. The more code there is that exercises a given API, the greater the cost of attempting to deprecate and replace that API. This doesn't affect the code we're writing today, but in general whenever someone is trying to evolve an API the default position is that all use-cases currently served by the existing API must be served by the new one. Having extra code that uses that API, even where we don't actually care about the use-case, restricts the way we can evolve the API by forcing us to justify why we don't want to support the feature.

All of this is a lengthy way of saying: code is a liability, not an asset. This is commonly misunderstood, because supported use cases are assets, and supported use-cases are enabled by code, so it can seem as though code is an asset. The important thing to realise is that the code itself is always a liability: it can just generate an asset that offsets its costs. This is why it's a bad idea to write code just because it would be easy to do so. In some sense, implementing simple-but-rare features is like eating junk food: it feels good at first, and doing so occasionally doesn't hurt much, but the long term consequences of frequently doing so are catastrophic.

In this particular instance, then, my reasoning goes like this:

1. This is new code, initial score is -1.
2. For a rare use-case, new score is -2.
3. This use-case can already be achieved today, albeit in an ugly way, so this does not add new _capabilites_ to NIO, new score is -2.5.
4. But the aesthetic is definitely better, new score is -1.

Better aesthetics are only ever worth it in cases where the use-case is either not supported at all or where we are saving huge chunks of code. Here we do neither. This is also why I'm ok with the other bits of code here: they are common paths, and having nice spellings for them is a quality of life improvement.

Reply:
I agree for the most part, however I think this is a slightly different case. Here we're providing a convenience API for setting socket options. In my mind this should be "all or nothing". If a TP developer is using the newer/cleaner API but is unable to find an option (e.g. `acceptConnections`) then they may assume we don't support it at all, causing confusion and/or frustration until they give up and use some other technology or spend significant time hunting through the NIO source.

Admittedly this can be resolved by adding some documentation, however:
- Which developer really jumps to documentation as their first port of call? (we're a lazy bunch)
- Code should be largely self-documenting (opinion)

Anyway, an interesting discussion for perhaps another time. Meanwhile I've removed :)

Reply:
I think a middle ground is to deprecate the older API and "newify" it somewhat, so that it's a little easier for other engineers to discover via code complete

Reply:
> I agree for the most part, however I think this is a slightly different case. Here we're providing a convenience API for setting socket options. In my mind this should be "all or nothing".

Sure, that makes sense, but then I have bad news, because the current `SocketOption` solution you're building on is not capable of representing all socket options. As @weissi has [hit upon here](https://github.com/apple/swift-nio/pull/1260#discussion_r348530477), the current SocketOption type cannot set generic value types for arbitrary option types, and is not type safe. This means it's a bad fit for "wide" socket options (such as the ones required for the various multicast groups) and it's a bad fit for type safety (because each option is not currently a separate type and so cannot communicate what type it wants).

On top of all of that, to be all-or-nothing requires a fairly complete understanding of what the available socket options are. As an example, [here](https://git.musl-libc.org/cgit/musl/tree/include/sys/socket.h#n199) is a list of the socket options supported by musl libc. There are at least 60 in the `SO_` space, plus many more in the [TCP namespace](https://git.musl-libc.org/cgit/musl/tree/include/netinet/tcp.h#n7), the [IP namespace](https://git.musl-libc.org/cgit/musl/tree/include/netinet/in.h#n178), as well as a couple for IPv6 and UDP.

To do a truly all-or-nothing solution requires a complete enumeration of all socket options the users may want to set, a platform-specific description of the type required by each one, and then sufficient code to get and set those. This then needs to be repeated on Linux with all relevant libc implementations as well as on the Apple platforms, including any version-specific annotations we might need to deal with there. We then need to also cover doing this in newer OS versions whenever new socket options come along, or whenever we add support for a new socket type (e.g. ethernet, raw sockets).

This is not necessarily a bad idea, to be clear: I just want to caution against the idea of doing an "all or nothing" socket option API that explicitly names each socket option. I don't think we want that: I think we want good names for the common options, and an escape hatch for everything else.